### PR TITLE
LUFA BootloaderHID

### PR DESCRIPTION
**Pull request by scott-t-wilson**
Updates the HID Bootloader with similar features as the DFU bootloader: QMK_LED/SPEAKER/ESC. Also adds a new target 'bootloaderhid' that triggers the new build.

### CODE REVIEW

**Comment by jackhumbert** [comment]
Nice! This is awesome :) is BootloaderHID compatible with the teensyloader? I'm not sure if we should be using the official VID/PIDs here.

I think it might make sense to commit some of these changes to @qmk/lufa, right? I'm not sure if committing them here will push them to our specific Lufa fork.

**Comment by scott-t-wilson** [comment]
Is qmk/lufa pulled in at some point? It doesn't look like its a subproject.

As far as teensyloader- I've been reading up on it, and apparently lufa was made 'incompatible' on purpose. Paul apparently requested that compatibility be obfuscated to make it harder on teensy clones. Its really easy to make it work, but I guess you shouldn't have a teensy compatible bootloader in qmk.

I do think a driverless bootloader is **extremely** important. Getting the DFU bootloader to work on windows 10 is a huge problem for a lot of people. I guess the best direction would be to have this HID bootloader, with QMK Toolbox compatibility added. I'm more than happy to open issues/pull requests on Toolbox when I get to that point, but I don't have a windows dev environment to test anything.

**Comment by jackhumbert** [comment]
Ah, it must be a subtree then. Let me look into this a bit - I'm not sure of the best way to get that updated then.

Yeah, I think we should develop a different VID/PID to use in coordination with the QMK Toolbox, to respect the work that PJRC's put into the Teensy system. I'm not sure if there's already a standard alternative.

Agreed! The solution we have lined-up for the Toolbox is to prompt them to install the drivers on start-up - I think this will be included in 0.0.10 for Windows at least. We'd be happy to support the BootloaderHID there as well :)

**Comment by harshitgoel96** [comment]
i somehow managed to delete bootloader on my ymd 96... can someone help me how to find/build one ? the default one was bootloaderhid i think.

**Comment by drashna** [comment]
@harshitgoel96 It may be best to open up a separate issue, since this seems unrelated to the PR.
That said, if you can find the bootloader HEX/BIN file for your chip, you can ISP flash it.

File: lib/lufa/Bootloaders/HID/Descriptors.c [hunk 1]
```diff
@@ -73,8 +73,8 @@ const USB_Descriptor_Device_t DeviceDescriptor =
 
 	.Endpoint0Size          = FIXED_CONTROL_ENDPOINT_SIZE,
 
-	.VendorID               = 0x03EB,
-	.ProductID              = 0x2067,
+	.VendorID               = 0x16C0,
```
Comment (drashna) [inline lib/lufa/Bootloaders/HID/Descriptors.c]:
```suggestion
	.VendorID               = 0x03EB,
```

File: lib/lufa/Bootloaders/HID/Descriptors.c [hunk 2]
```diff
@@ -73,8 +73,8 @@ const USB_Descriptor_Device_t DeviceDescriptor =
...
+	.ProductID              = 0x0478,
```
Comment (drashna) [inline lib/lufa/Bootloaders/HID/Descriptors.c]:
```suggestion
	.ProductID              = 0x2067,
```

**Comment by drashna** [comment]
@scott-t-wilson Could you fix the device IDs, so we can see about merging this?

File: tmk_core/avr.mk [hunk 3]
```diff
@@ -304,6 +304,20 @@ bootloader:
 	printf "BootloaderDFU.hex copied to $(TARGET)_bootloader.hex\n"
 	cp lib/lufa/Bootloaders/DFU/BootloaderDFU.hex $(TARGET)_bootloader.hex
 
+bootloaderhid:
+	make -C lib/lufa/Bootloaders/HID/ clean
+	echo "#ifndef QMK_KEYBOARD\n#define QMK_KEYBOARD\n" > lib/lufa/Bootloaders/HID/Keyboard.h
+	echo `grep "MANUFACTURER" $(ALL_CONFIGS) -h | tail -1` >> lib/lufa/Bootloaders/HID/Keyboard.h
```
Comment (yanfali) [inline tmk_core/avr.mk]: Heredoc conversion would be helpful

**Comment by drashna** [comment]
@scott-t-wilson would you be willing to revisit/update this PR? 

The lufa fork in question is up to date now, and there is a PR to switch to using it as a submodule (#6245)

**Comment by drashna** [comment]
@scott-t-wilson this PR is basically broken now.  LUFA has been moved to a submodule, so it's in its own repo now.  

If you could, move the relevant changes there, and update this PR.

**Pull request closed without merging**